### DISCUSSION TITLE
Pokemon Emerald RTC

### DISCUSSION THREAD

**Discussion by cangelo98** [kwsch/PKHeX#3771] [Q&A]
Hello all, I am having an issue adjusting the rtc for my Emerald cart to get all of the time based events to work. I tried enabling the change rtc event from the event flags tab. On the title screen I press left+select+b and it takes me to the menu to adjust the rtc. I can't seem to figure out how it works though. It shows previous time and present time but on pkhex the rtc feature shows intial time and time elapsed. they don't seem to have anything in common because when I change the time in pkhex it doesen't seem to match what I put in game. I'll send a picture to show what I mean. Sometimes I can get the berries to grow or get the Lilycove lottery to work by playing around with it enough but I can't get the man in pacifidlog town to give me the tm every 7 days. He wants me to wait 537 days. Sometime I get it to 8 days but when the next day rolls over he still wants me to wait 8 days. I can't seem to find much documentation on how the rtc works in r/s/e so if someone can give me some insight it would be much appreciated.
![IMG-2115](https://user-images.githubusercontent.com/124932562/217936323-bfce282f-d3a3-4ec7-b540-ecccb6412825.jpg)

**Comment by kwsch** [comment]
Try applying the Berry Fix. The game has issues dealing with dates 1-2 years after release, which your values might be within.
https://bulbapedia.bulbagarden.net/wiki/Berry_glitch

https://github.com/kwsch/PKHeX/blob/3e6a036c549c3b2fbb5cbea3a3d9806f5c07b215/PKHeX.Core/Saves/SAV3E.cs#L62-L72
https://github.com/pret/pokeemerald/blob/6d6a876dbdc0d0e5b89d3b586dd8a8d7e0ce7108/include/global.h#L509-L510
https://github.com/pret/pokeemerald/blob/79c5a5983e4bd4cffd39c1c6ef32d66fc63f516d/src/reset_rtc_screen.c#L581-L601

Here's an example RTC you can try copying:
![image](https://user-images.githubusercontent.com/6393368/217982949-30d42688-a008-4bb9-96b6-179b4a6fae9b.png)

**Reply by cangelo98** [reply]
Okay I will give it a shot. I wanted to ask before I start though which column do I put the current time into, would it be time elapsed or initial time? Also does the in game rtc settings differ from what is shown on pkhex because they don't seem to work the same?

**Comment by cangelo98** [comment]
Update:
![IMG-2169](https://user-images.githubusercontent.com/124932562/225764835-ef96f38e-dfad-4509-add0-050104b1deb0.jpg)
 My RTC is fully working. Sorry it took a while for an update, I've been busy. I am curious though on pkhex it shows that the initial time is 57095 days. I have no idea what that means because it doesn't match any of the values on the reset rct screen. Can anyone please explain that?

**Reply by kwsch** [reply]
57095 is -8,441 if we treat the u16 as a s16 instead. Seems close to your date value.

Not sure why it's negative, but it is.

**Reply by Lorenzooone** [reply]
As a heads up: you're reading the offset which is applied to the battery's internal time, with a (s16) subtraction, to get the value stored in the battery to the one requested by the user when it is first set. This is constant, and won't change once it's applied. The battery time will increase though, and the DAYS var with it. That is what increases the time.

So... To add a day, you need to subtract a day from the initial time.

Also: Day 8445 is in the 23rd year since 2000, so it's basically up to date.

**Reply by profbinks** [reply]
Can someone explain this a little better? u16? s16? What exactly needs to be done to get a fully working RTC after the battery is replaced with an old save??

**Reply by profbinks** [reply]
Also @cangelo98, curious to see what happens June 18, 2027. According to the post date here, you will hit 10,000 days on that date and the internal game clock is only capable of handling 9999 days. I'm wondering how screwed up it will be at that point. It should start over at 1 in game but then you're rolling back in time and will more than likely have all the same issues again.

**Reply by Lorenzooone** [reply]
Just use the RTC reset menu and set it to whatever value you want (1 is a valid one, for example).
The game will be fine and handle setting everything on its own.
This is not something that you should worry about.

Issues may arise if you don't use the RTC reset menu. But as long as you use it, you will be fine.

Issues may (keyword may, not guaranteed) arise in 2089, however it's not that important, as the year value isn't tracked, and any battery you have will likely be dead by then, meaning you will be able to make this work by resetting the internal RTC chip year to 0 or something (which should happen when you change a battery).

**Reply by profbinks** [reply]
@Lorenzooone Yes you can reset the RTC but in my testing and from reading the experiences of others, simply resetting the RTC via PKHeX does not make all time-based events work in the game again. Obviously jumping time will make any planted berries disappear. Aside from that and as OP mentioned, NPC behavior no longer functions as expected as certain values (which I have yet to determine) use eachother as reference points. Restart the save file and all issues are solved, but if you're using the same save file it isn't as simple as just reseting the RTC values in PKHeX.

**Reply by Lorenzooone** [reply]
Yes. Which is why I said to use the RTC reset menu, which is a special menu accessible in Emerald.

Idk about PKHex's option, it may be missing either updating the Days VAR or the Berry Time.

However, the game's menu will make this work.

The main issue is that PKHex cannot know in what state the cart's RTC chip is, so it may not be able to make a really proper change.
The game actually can read the chip, so it will be fine.

**Reply by profbinks** [reply]
This only allows you to change the reference timestamp for when your clock was intially set in game 
![image](https://github.com/kwsch/PKHeX/assets/17938291/04f7dd4c-5b57-4386-b327-f10eefb70107)
You cannot change the current time that the game clock is running unless you use PKHeX

**Reply by profbinks** [reply]
Hmm it changes them both when you update it in game, I think a big part of the problem here is trying to use the RTC reset menu in PKHeX in conjunction with the in game menu as they function differently.

**Reply by Lorenzooone** [reply]
Emerald changes both the reference timestamp and the berry time (which is the time when your berries where last updated, and what PKHex calls "elapsed time"), but it also resets the variables properly to make sure everything works, which, from what you're saying, PKHex doesn't (though I haven't confirmed it myself: you could be using it wrong, as you're likely setting both the berry time and the regular time to a positive, when one should be a negative).
Emerald never touches the RTC chip. Unless it failed (and even then, it only resets it).
PKHex does not allow changing the current time. It also works on that reference timestamp (and the berry time if you also edit it), as it has no access to the RTC chip.

To change the RTC Chip's actual time, you need to build and use a ROM like the one I have in this branch: https://github.com/Lorenzooone/Pokemon-Gen3-to-Gen-X/tree/time_setter

This all seems to stem from a lack of understanding of how the RTC code works in Emerald and what each value does. So I'd suggest you read that previous link, this one https://github.com/Lorenzooone/Pokemon-Gen3-to-Gen-X/blob/main/source/gen3_clock_events.c , this one https://github.com/pret/pokeemerald/blob/master/src/siirtc.c and this one https://github.com/pret/pokeemerald/blob/master/src/rtc.c .

TLDR; If you don't know what you're doing, use the In-game RTC reset menu, it will work fine.

**Reply by profbinks** [reply]
So if you reset the RTC in game each time the battery is replaced, an old save file will always automatically recalibrate? Will NPC wait times be screwed up?

**Reply by Lorenzooone** [reply]
Yes. There may be a full day of delay, but it should work fine after that. Iirc, it should also preserve your berries, assuming strange stuff hasn't happened and they didn't get reset beforehand (which may happen due to the way the clock works when the battery dies).

**Reply by profbinks** [reply]
Alright thanks, I'll give it a shot and report back if I see any more issues.

**Reply by profbinks** [reply]
It also sounds like it would never work moving the save file back and forth between cart and emulator since the emulator updates the time elapsed/berry time to match your pc system time in reference to 1/1/2000.

**Reply by MrPerson0** [reply]
> curious to see what happens June 18, 2027. According to the post date here, you will hit 10,000 days on that date and the internal game clock is only capable of handling 9999 days. I'm wondering how screwed up it will be at that point. It should start over at 1 in game but then you're rolling back in time and will more than likely have all the same issues again.

FWIW, I tested it out by using RTCread to change the date on my Emerald cartridge. As mentioned, there was no issue with moving it to 2027 since the game seems to keep track of the fifth digit (in terms of days passed), and daily events worked like usual. As @Lorenzooone mentioned, moving the year past 2089 is when daily events no longer started to occur, probably due to a bug in the RTC. It really isn't a big deal since the RTC rolls back to 2000 after 2099, AFAIK.

That being said, from what I understand, the in-game clock reset mainly changes the "Time Elapsed" value that you see in PKHeX's Real Time Clock Editor. The main thing you want to note is, as long as the Days counter in "Time Elapsed" (in the same RTC editor) surpasses the "Days passed" value in Event Flags > Event Constants, daily events should work once again. 

This is why using "Reset RTC" in PKHeX usually isn't enough to fix daily events after you replace a battery. You need to change the RTC clock itself, using GBA or DS homebrew and make sure the days surpass your "Days passed" value. Emulators do this automatically since they match your system clock, and why people recommend changing your cart's RTC to the current date and time.

>It also sounds like it would never work moving the save file back and forth between cart and emulator since the emulator updates the time elapsed/berry time to match your pc system time in reference to 1/1/2000.

As long as your cart's RTC matches today's date and time, you can move between cart and emulator with no hassle in regards to the RTC.

**Reply by profbinks** [reply]
The in game RTC reset (not RTCRead) only goes up to 9999 days, so after 2027 you can no longer accurately set the RTC on the cart to allow hot swapping if I understand correctly? The game code would need to be changed another way, either PKHeX or RTCRead but you'd also have to know exactly what to change, because it would be multiple values

**Reply by Lorenzooone** [reply]
Assuming the RTC Chip's time is correct (do note that it does drift by a few seconds every year or so), it would work fine. Just setting your day to 1 or whatever will work. It's the offset you're editing. Not the chip's time.

**Reply by MrPerson0** [reply]
Even if the ingame RTC reset itself doesn't read above 9999 days, PKHeX does: 
![image](https://github.com/kwsch/PKHeX/assets/26208088/328057b0-e4d0-402a-af56-c75051b92213)

Also, RTCRead is a GBA Homebrew app that allows you to edit your GBA cart's RTC. You should use that or the DS variant to edit yours.

**Reply by profbinks** [reply]
Can you do this without a DS? I'm just using an OG GBA

**Comment by profbinks** [comment]
@kwsch the berry glitch isn't relevant here as OP is playing emerald. Nintendo fixed this issue in Emerald which is why Emerald can send the patch over to Ruby and Sapphire via link cable.

**Comment by profbinks** [comment]
> Assuming the RTC Chip's time is correct (do note that it does drift by a few seconds every year or so), it would work fine. Just setting your day to 1 or whatever will work. It's the offset you're editing. Not the chip's time.

Is this in reference to using with an emulator? All that matters is the time and not days? I know playing on an emulator updates the elapsed time in reference to days past 1/1/2000

**Reply by profbinks** [reply]
This is how it saved when saving on my emulator. It's off by 365 days but the time is based on the PC system time
![image](https://github.com/kwsch/PKHeX/assets/17938291/d9113890-e6ab-4da7-b6f6-942ae529f6d8)

**Reply by Lorenzooone** [reply]
With time, I meant the date as well, sorry.
If the actual RTC Chip (not the offset)'s date and time are correct, it should work fine. So you can use some homebrew to change that.

Regardless of that, there was talk at one point about getting mgba to properly handle different RTC dates + times depending on the game, which would also fix the risk of the 2089 issue, though idk how that went.

**Reply by MrPerson0** [reply]
>It's off by 365

Is this a Ruby/Sapphire cart? If so, it being a year off is normal due to the Berry Glitch.

>Regardless of that, there was talk at one point about getting mgba to properly handle different RTC dates + times depending on the game, which would also fix the risk of the 2089 issue, though idk how that went.

Ideally, they would allow people to edit it as they choose.

**Reply by profbinks** [reply]
Yes I have been doing all my testing with a Sapphire cart and the playback software for the GB operator. So it sounds like the most reliable way to ensure date and time stay as closely synced as possible is using RTCread on a GBA flashcart? Not familiar with it so I'll have to look into it. And one other point of clarification, which value is the offset? It's hard to follow when they're labeled differently everywhere.

**Reply by MrPerson0** [reply]
>So it sounds like the most reliable way to ensure date and time stay as closely synced as possible is using RTCread on a GBA flashcart?

Yep, that's the idea. 

Another option is if you can get the GBxCart RW. That can write to the RTC of the game cartridge by reading the data from the save file (mGBA is the only emulator that appends RTC info to the save file, AFAIK). Since mGBA gets its time from the system clock, every time you write back to the cartridge, the time would be synced to the system clock. This is somewhat redundant since you already have the GB operator, but I don't think that device writes or reads the RTC from a cartridge.

**Reply by profbinks** [reply]
Hmm. yeah I don't reallly want to buy another piece of hardware that does pretty much the same thing, if possible I'd like to try the RTCRead solution, just not sure exaclty how to do that, or if it can be done without having a DS lite

**Reply by MrPerson0** [reply]
AFAIK, the only way to load GBA homebrew is through the use of a GBA flash cart, so you're required to purchase extra hardware.

**Reply by profbinks** [reply]
I have the EZ Flash Omega DE, will that work?

**Reply by Lorenzooone** [reply]
Assuming that is a multibootable (I think it should be), you could use a Wii and a GCN-GBA Link  Adapter to launch the homebrew.

**Reply by Lorenzooone** [reply]
Yes. The EZ Flash will work fine. Just load that ROM in there.

**Reply by profbinks** [reply]
I've got a GBA sp and a link cable?

**Reply by MrPerson0** [reply]
Yep, that flash cart should work. You can look into the homebrew software from here: http://www.furlocks-forest.net/wiki/?page=Pokemon_Ruby/Sapphire_New_Battery_Glitch

**Reply by profbinks** [reply]
I think I already downloaded the zip from one of the links there, is it just a matter of putting this rom on the sd card and into the flashcart? 
![image](https://github.com/kwsch/PKHeX/assets/17938291/71cf229f-2b40-443b-980b-6ef9d9afb042)

**Reply by MrPerson0** [reply]
> is it just a matter of putting this rom on the sd card and into the flashcart?

Exactly. Once you have loaded the file, you can take out your cartridge and put your RSE cartridge in there. You press start to refresh and check the current RTC, then you can press select to edit the RTC.

**Reply by profbinks** [reply]
Does it work the same when linking two GBAs? or how does that work?

**Reply by MrPerson0** [reply]
You just need one GBA. Once you load RTCread, you can take out the cartridge and replace it with an RSE one with no issue.

**Reply by profbinks** [reply]
Holy crap it's done 
![image](https://github.com/kwsch/PKHeX/assets/17938291/1db142d2-e414-4d60-9472-d47e6d9a6fe8)
Just need to see how it functions in game now. This is how pkhex reads it after setting with RTCRead to current day and time, not sure I understand it, but the UI is better for RTCRead

**Reply by profbinks** [reply]
@MrPerson0 also the "days passed" constant is the same day value as the time elapsed but I'm assuming since the time elapsed also has hours, minutes, seconds, it's technically ahead and days passed is considered lower?

**Reply by profbinks** [reply]
This is the save after loading it into the emulator, so it still appears to be messed up 
![image](https://github.com/kwsch/PKHeX/assets/17938291/7651eac4-a671-411a-913a-34306a80f94f)

**Reply by Lorenzooone** [reply]
When using RTCRead, did you swap the day with the month? I think that may be what is happening. (Either that, or the emulator has that issue. Idk)

**Reply by MrPerson0** [reply]
>I'm assuming since the time elapsed also has hours, minutes, seconds, it's technically ahead and days passed is considered lower?

From what I have seen, the "days passed" constant is just a check for daily events (maybe the weekly event in Pacifidlog Town too?), so it doesn't care about hours/minutes/seconds elapsed. Only thing it looks for is if the Time Elapsed moves on to the next day, that's when the "days passed" changes as well.

Also, your initial time definitely looks off. Usually, when I replace the battery on a cart (that used to have a stock battery), I will first do click on Reset RTC so the initial time is set to 0. That way, the current time is set to whatever time is on the new RTC so you can easily edit it with RTCread. However, this might kill off any berries you had planted in the overworld.

>When using RTCRead, did you swap the day with the month?

Yeah, that's another thing to note. With RTCread, the first number is the day of the month, the second number is the actual month.

**Reply by profbinks** [reply]
@MrPerson0 you reset it with the in-game menu first? to 0 days 0 hrs? or set the clock?

**Reply by MrPerson0** [reply]
No, I just reset it with PKHeX using the "Reset RTC" button. Then, I set the cartridge's clock with RTCread to the current date and time, and from there, everything is good to go.

**Reply by profbinks** [reply]
Looks to be working properly 
![image](https://github.com/kwsch/PKHeX/assets/17938291/b81000b7-a133-4a55-8acc-0208b6281f27)
Time will tell

**Reply by profbinks** [reply]
@MrPerson0 @Lorenzooone Thank you both for all your input and timely responses. This level of documentation/troubleshooting support is not something I've been able to find anywhere! Hopefully this helps others who are stuck with the same issue as well.

**Comment by jonasknobloch** [comment]
> However, this might kill off any berries you had planted in the overworld.

@MrPerson0 do you have an idea why that happens?

I was playing around with an Everdrive trying to match its RTC with the current real world time. After updating the RTC, planted berries just vanished. When setting the RTC to a time between "Initial Time" and "Time Elapsed", freshly planted berries became fully grown instantly.

I was wondering what would happen if I set the RTC to a time slightly before "Initial Time" + "Time Elapsed", with the game running when that boundary is crossed. After saving and restarting the berries are gone again. But interestingly if I go back to a time before "Time Elapsed" or disable the RTC altogether, the fully grown berries are back. Doing the same without crossing the boundary in-game, leads to freshly planted berries.

**Reply by jonasknobloch** [reply]
![Screenshot 2024-03-21 115700](https://github.com/kwsch/PKHeX/assets/7591950/e8a55171-80a9-4e53-9b29-bf53854acfe4)

![Screenshot 2024-03-21 115735](https://github.com/kwsch/PKHeX/assets/7591950/693432b8-a30e-452b-bf3f-6fb744ed03db)

Is all that intended beahviour? Meaning the berries vanish because they have exceeded their maximum regrowth count?

**Reply by MrPerson0** [reply]
> do you have an idea why that happens?

I believe it's simply due to the new elapsed days being significantly higher than it was before. On a normal save, the elapsed days might hover around 1000 days once the save battery dies out. Once you put in a new battery, set the game's initial time to zero, and the battery's RTC to the current time, the elapsed days will be much higher (8000 or so now). According to Bulbapedia, berries seem to regrow 9 times, so the new day difference is much longer than the growth cycle: https://bulbapedia.bulbagarden.net/wiki/Berry#Growth_and_care

As for your issue with the berries, it might not be documented, but I guess the difference in between the days elapsed is because you didn't reach the maximum regrowth count yet, which would be 9.
